### ISSUE TITLE
Apparent Floating Point Error

### ISSUE THREAD

**Issue by alexk101** [google/jax#15646]
### Description

Hello,
I have been working on optimizing some routines for some machine learning methods we would like to apply to our work, but I seem to be getting some difference between the original numpy version of this code, and my, likely, semi optimized version in Jax. I attach an example which is reproducible on my machine below. If I am missing something somewhere, I apologize, this is my first pass at using Jax

```python
import jax
from jax import random, vmap, jit
import jax.numpy as jnp
import numpy as np

# Numpy
def np_computeMeanVec(X, y, n):
    mean_vectors = []
    for cl in n:
        mask = y==cl
        print(mask.shape)
        mean_vectors.append(np.mean(X[mask], axis=0))
    return np.stack(mean_vectors)

# JAX
def get_mask(y: jnp.ndarray, n: int):
    mask = y==n
    return mask.flatten()


def computeMeanVec(X: jnp.ndarray, mask: jnp.ndarray) -> jnp.ndarray:
    """Compues the d-dimensional mean vectors for the class defined by mask

    Args:
        X (jnp.ndarray): Feature matrix
        mask (jnp.ndarray): Mask defining class membership

    Returns:
        jnp.ndarray: mean feature vector for class defined by mask
    """
    return jnp.mean(X, where=mask.reshape(-1,1), axis=0)


vmapMask = jit(vmap(get_mask, in_axes=(None,0)))
vmapCMV = jit(vmap(computeMeanVec, in_axes=(None,1)))


def computeAllMeanVec(X, y, n):
    mask = vmapMask(y.reshape(-1,1), n).T
    print(mask.shape)
    return vmapCMV(X,mask)


if __name__ == "__main__":
    key = jax.random.PRNGKey(42)

    # offending float is 4.2950333e+09
    X = np.full((400,2), 4.2950333e+09).astype(np.float32)
    y = np.full((1,400), 2)[0].astype(np.int32)
    n = np.unique(y).astype(np.int32)

    # Reference
    mean_org = np_computeMeanVec(X, y, n)

    X = jnp.array(X, dtype=jnp.float32)
    y = jnp.array(y, dtype=jnp.int32)
    n = jnp.array(n, dtype=jnp.int32)
    # JAX Optimized
    mean_vectors = computeAllMeanVec(X, y, n)

    # err -> [[-24064. -24064.]]
    err = mean_vectors - mean_org    
    print(err)
```

Thanks, Alex

### What jax/jaxlib version are you using?

jax v0.4.8

### Which accelerator(s) are you using?

GPU

### Additional system info

Python 3.11.3, Ubuntu 22.04

### NVIDIA GPU info

```
Mon Apr 17 22:09:24 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3070 T...    On | 00000000:01:00.0 Off |                  N/A |
| N/A   43C    P8               17W / 115W|    248MiB /  8192MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      2786      G   /usr/bin/gnome-shell                        245MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by jakevdp** [comment]
Hi - thanks for the question. It's important here to look at not just the difference, but also the relative difference:
```python
print(mean_org)  # numpy result
# [[4.295057e+09 4.295057e+09]]

print(mean_vectors)  # jax result
# [[4.2950333e+09 4.2950333e+09]]

err = mean_vectors - mean_org    
print(err)  # difference
# [[-23552. -23552.]]

print(err / mean_vectors)  # relative difference
# [[-5.483543e-06 -5.483543e-06]]
```
Your numpy and JAX results differ by about 1 part in $10^7$. You're using float32 values, and float32 has a 23-bit mantissa, meaning that outputs will normally have errors on order of 1 part in $2^{23}$, which is roughly 1 part in $10^7$. So it looks like your output is consistent with the normal roundoff error of floating point operations, and therefore is probably working as expected.

**Comment by jakevdp** [comment]
Thanks - another note: if your computation requires more accuracy, you could [enable 64-bit precision](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision) and then you'll have several orders of magnitude more precision to work with.
